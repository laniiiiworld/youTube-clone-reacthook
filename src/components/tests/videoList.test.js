import '@testing-library/jest-dom';
import { render, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react';
import { Route } from 'react-router-dom';
import { withAllContexts, withRouter } from '../../tests/utils';
import { fakeRelatedVideos, fakeSearchVideos, fakeVideos } from '../../tests/videos';
import VideoList from '../videoList/videoList';

describe('VideoList', () => {
  const fackYouTube = {
    search: jest.fn(),
  };

  beforeEach(() => {
    fackYouTube.search.mockImplementation((keyword, videoId) => {
      if (videoId) return fakeRelatedVideos;
      return keyword ? fakeSearchVideos : fakeVideos;
    });
  });

  afterEach(() => {
    fackYouTube.search.mockReset();
  });

  it('renders correctly', async () => {
    const { asFragment } = renderWithPath('/');
    await waitForElementToBeRemoved(() => screen.getByTitle('loading'));
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders all videos when keyword is not specified', async () => {
    renderWithPath('/');
    expect(fackYouTube.search).toHaveBeenCalledWith(undefined, undefined);
    await waitFor(() => expect(screen.getAllByRole('listitem')).toHaveLength(fakeVideos.length));
  });

  it('when keyword is specified, renders search results', async () => {
    const searchKeyword = 'fake-keyword';
    renderWithPath(`/search/${searchKeyword}`);

    expect(fackYouTube.search).toHaveBeenCalledWith(searchKeyword, undefined);
    await waitFor(() => expect(screen.getAllByRole('listitem')).toHaveLength(fakeSearchVideos.length));
  });

  it('renders related videos correctly', async () => {
    const relatedVideoId = 'related-video-id';
    renderWithPath(`/detail/${relatedVideoId}`);

    await waitFor(() => expect(screen.getAllByRole('listitem')).toHaveLength(fakeRelatedVideos.length));
  });

  it('renders loading state when items are being fetched', () => {
    renderWithPath('/');

    expect(screen.getByTitle('loading')).toBeInTheDocument();
  });

  it('renders error state when fetching items fails', async () => {
    fackYouTube.search.mockImplementation(async () => {
      throw new Error('error');
    });
    renderWithPath('/');

    await waitFor(() => expect(screen.getByText(/Error/i)).toBeInTheDocument());
  });

  function renderWithPath(path) {
    return render(
      withAllContexts(
        withRouter(
          //
          <>
            <Route path='/' element={<VideoList />} />
            <Route path='/search/:keyword' element={<VideoList />} />
            <Route path='/detail/:videoId' element={<VideoList videoId='videoId' />} />
          </>,
          path
        ),
        fackYouTube
      )
    );
  }
});
