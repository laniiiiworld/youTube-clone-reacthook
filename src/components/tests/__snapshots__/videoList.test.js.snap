// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoList renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="videoList grid"
  >
    <li
      class="videoItem grid"
      data-testid="video-list"
    >
      <div
        class="thumbnail"
      >
        <img
          alt="title4"
          class="thumbnailImg"
          src="http://image4/"
        />
      </div>
      <div
        class="info"
      >
        <p
          class="title"
        >
          title4
        </p>
        <p
          class="channelTitle"
        >
          channelTitle4
        </p>
        <p
          class="publishedAt"
        >
          just now
        </p>
      </div>
    </li>
    <li
      class="videoItem grid"
      data-testid="video-list"
    >
      <div
        class="thumbnail"
      >
        <img
          alt="title2"
          class="thumbnailImg"
          src="http://image2/"
        />
      </div>
      <div
        class="info"
      >
        <p
          class="title"
        >
          title2
        </p>
        <p
          class="channelTitle"
        >
          channelTitle2
        </p>
        <p
          class="publishedAt"
        >
          just now
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
